### PR TITLE
Sign up/sign in

### PR DESCRIPTION
**Pull request by aniansson**
## Pivotal Tracker story

https://www.pivotaltracker.com/story/show/139265229

## Changes proposed in this pull request:

This feature makes it possible for new users to sign up + existing users to sign in

## What I have learned working on this feature:

How everything links together in the client  (modals, views, controllers, scope etc.) and how to create not only a sign in option but sign up in a similar way. And how it links to the API.

### Ready for review!

### CODE REVIEW

File: www/js/controllers.js [hunk 1]
```diff
@@ -1,56 +1,131 @@
...
+    // Form data for the login modal
+    $scope.loginData = {};
+
+    $scope.signupData = {};
+
+    // Create the login modal that we will use later
+    $ionicModal.fromTemplateUrl('templates/login.html', {
+      scope: $scope
+    }).then(function (modal) {
+      $scope.loginmodal = modal;
```
Comment (tochman) [inline www/js/controllers.js]: I would use pascalCase here: `loginModal`. Same thing below.

File: www/js/controllers.js [hunk 2]
```diff
@@ -1,56 +1,131 @@
...
+          $scope.closeSignup();
+          $ionicLoading.show({
+            template: 'Successfully signed up!',
+            duration: 1000
+          });
+        })
+        .catch(function (response) {
+          $ionicLoading.hide();
+          $scope.errorMessage = response;
+          $scope.showAlert('Failure', response.data.errors.full_messages);
```
Comment (tochman) [inline www/js/controllers.js:64]: `errors.full_messages` is an Array. I would like to see a screenshot of what this looks like if there are more than one error

File: www/js/controllers.js [hunk 3]
```diff
@@ -1,56 +1,131 @@
...
-    { title: 'Cowbell', id: 6 }
-  ];
-})
-
-.controller('PlaylistCtrl', function($scope, $stateParams) {
-});
+
+    $rootScope.$on('auth:login-success', function (ev, user) {
+      $scope.currentUser = angular.extend(user, $auth.retrieveData('auth_headers'));
+    });
```
Comment (tochman) [inline www/js/controllers.js:107]: What does this do for you?

File: www/js/controllers.js [hunk 4]
```diff
@@ -1,56 +1,131 @@
...
+
+    $rootScope.$on('auth:registration-email-success', function (ev, user) {
+      console.log("hi");
+      $scope.currentUser = angular.extend(user, $auth.retrieveData('auth_headers'));
+    });
+  })
```
Comment (tochman) [inline www/js/controllers.js]: Same question here, why are you including this?

File: www/js/controllers.js [hunk 5]
```diff
@@ -1,56 +1,131 @@
...
+
+
+  .controller('dishesCtrl', function ($state, $scope, dishes, $ionicLoading, $ionicPopup) {
+    $scope.$on('$ionicView.enter', function () {
+      $ionicLoading.show({
+        template: 'Retrieving delicious food'
+      });
+      dishes.query(function (response) {
```
Comment (tochman) [inline www/js/controllers.js]: Is `dishes` a name of a service? I would revisit this naming. I usually use singular format and capital. `Dish` Like a class name in ruby. But there are many standards. `dishService` is one option.

File: www/js/controllers.js [hunk 6]
```diff
@@ -1,56 +1,131 @@
...
+        $scope.dishCollection = response.entries;
+        $ionicLoading.hide();
+      }, function (response) {
+        //error
+      });
```
Comment (tochman) [inline www/js/controllers.js]: If you dont intend to handle errors, plz remove this callback.

**Review by tochman — CHANGES_REQUESTED** [review]

File: www/js/controllers.js (hunk 1 above)
Comment (Philippoes) [reply www/js/controllers.js]: Got it.

File: www/js/controllers.js (hunk 2 above)
Comment (Philippoes) [reply www/js/controllers.js:64]: screen shot 2017-02-09 at 19 36 06

File: www/js/controllers.js (hunk 3 above)
Comment (Philippoes) [reply www/js/controllers.js:107]: I assumed this was an AUTH event. Like when the login/signup is successful, currentUser becomes the user which has just logged in. It is basically a way for the app to know which user is currently using it. And I think that when there's a new request (for example add to order where user has to be logged in), the app will send that header with the token of the user currently logged in with the request, so the api knows which user's order to handle. Am I right? @tochman

File: www/js/app.js [hunk 7]
```diff
@@ -4,42 +4,61 @@
...
-
-    }
-    if (window.StatusBar) {
-      // org.apache.cordova.statusbar required
-      StatusBar.styleDefault();
-    }
-  });
-})
+angular.module('homechef', ['ionic', 'homechef.controllers', 'homechef.services', 'ng-token-auth', 'ngResource'])
+  .constant('API_URL', 'http://localhost:3000/api/v1')
```
Comment (diraulo) [inline www/js/app.js]: shouldn't this URL be the one deployed on heroku?

File: www/js/controllers.js (hunk 2 above)
Comment (tochman) [reply www/js/controllers.js:64]: The formatting feels a bit off. Are you guys happy with this?

File: www/js/controllers.js (hunk 3 above)
Comment (tochman) [reply www/js/controllers.js:107]: With ngTokenAuth you automatically have 
$scope.user available. No need for any further setup. Try without this.

**Review by diraulo — CHANGES_REQUESTED** [review]
see comment about api url

File: www/js/app.js (hunk 7 above)
Comment (aniansson) [reply www/js/app.js]: We had problems with the deployed one therefore we put int he localhost one

File: www/js/controllers.js (hunk 3 above)
Comment (Philippoes) [reply www/js/controllers.js:107]: I tried without this. Now I cannot access the current user for its email or checking if it's there. I have googled on how to use the ng token auth default user helper but with no luck.

File: www/js/app.js (hunk 7 above)
Comment (Philippoes) [reply www/js/app.js]: It is changed to the heroku app. now. We used it mainly for testing purposes.

**Review by tochman — APPROVED** [review]

File: www/js/controllers.js (hunk 3 above)
Comment (tochman) [reply www/js/controllers.js:107]: You can probably access `$scope.user` without it.

Comment (Philippoes) [reply www/js/controllers.js:107]: Okay well. I tried setting it: `<ion-item menu-close ng-if="!user" ng-click="login()">` (see ng-if) and the menu item still shows up even though i logged in. I tried console logging user.email but with no result.  @tochman

**Review by diraulo — APPROVED** [review]

**Pull request merged**